cellfinder/core/detect/filters/setup_filters.py

```diff
@@ -1,70 +1,425 @@
-
-
-def get_cell_detector(
-    *, plane_shape: Tuple[int, int], ball_z_size: int, z_offset: int = 0
-) -> CellDetector:
-    plane_height, plane_width = plane_shape
-    start_z = z_offset + int(math.floor(ball_z_size / 2))
-    return CellDetector(plane_width, plane_height, start_z=start_z)
-
-
-def setup_tile_filtering(plane: np.ndarray) -> Tuple[int, int]:
-    """
-    Setup values that are used to threshold the plane during 2D filtering.
-
-    Returns
-    -------
-    clipping_value :
-        Upper value used to clip planes before 2D filtering. This is chosen
-        to leave two numbers left that can later be used to mark bright points
-        during the 2D and 3D filtering stages.
-    threshold_value :
-        Value used to mark bright pixels after 2D filtering.
-    """
-    max_value = get_max_possible_value(plane)
-    clipping_value = max_value - 2
-    thrsh_val = max_value - 1
-
-    return clipping_value, thrsh_val
+
+MAX_TORCH_COMP_THREADS = 12
+# As seen in the benchmarks in the original PR, when running on CPU using
+# more than ~12 cores it starts to result in slowdowns. So limit to this
+# many cores when doing computational work (e.g. torch.functional.Conv2D).
+#
+# This prevents thread contention.
+
+
+@dataclass
+class DetectionSettings:
+    """
+    Configuration class with all the parameters used during 2d and 3d filtering
+    and structure splitting.
+    """
+
+    plane_original_np_dtype: Type[np.number] = np.uint16
+    """
+    The numpy data type of the input data that will be passed to the filtering
+    pipeline.
+
+    Throughout filtering at key stages, the data range is kept such
+    that we can convert the data back to this data type without having to
+    scale. I.e. the min/max of the data fits in this data type.
+
+    Except for the cell detection stage, in that stage the data range can be
+    larger because the values are cell IDs and not intensity data anymore.
+
+    During structure splitting, we do just 3d filtering/cell detection. This is
+    again the data type used as input to the filtering.
+    """
+
+    detection_dtype: Type[np.number] = np.uint64
+    """
+    The numpy data type that the cell detection code expects our filtered
+    data to be in.
+
+    After filtering, where the voxels are intensity values, we pass the data
+    to cell detection where the voxels turns into cell IDs. So the data type
+    needs to be large enough to support the number of cells in the data.
+
+    To get the data from the filtering data type to the detection data type
+    use `detection_data_converter_func`.
+
+    Defaults to `uint64`.
+    """
+
+    plane_shape: Tuple[int, int] = (1, 1)
+    """
+    The shape of each plane of the input data as (height, width) - i.e.
+    (axis 1, axis 2) in the z-stack where z is the first axis.
+    """
+
+    start_plane: int = 0
+    """The index of first plane to process, in the input data (inclusive)."""
+
+    end_plane: int = 1
```

Comment:
The intention here is the values the variables take are fully valid - they are just defaults for when you don't specify all during testing etc. The `-1` default is (still) used in `main` where it is be translated to the max plane when the settings is instantiated.

So we don't use `-1` here because it's not a valid slice (ignoring negative slicing). And `1` should be valid because there should be at least one plane.